### ISSUE TITLE
Breaks Checkout with Klarna - Allows payment to be taken but stops order being created in Magento.

### ISSUE THREAD

**Issue by TomJ1588** [dotmailer/dotmailer-magento2-extension-sms#7]
Hello,
Found out over the weekend, that the SMS extension here. (Which we have disabled in config) has caused an error - where a customer completed checkout, payment was taken, but this throws an exception and stopped Magento from saving the order.

Magento                                                                            2.4.6-p2
dotdigital/dotdigital-magento2-os-package                         2.2.2
dotdigital/dotdigital-magento2-extension-sms                    1.7.1
klarna/m2-klarna                                                                  2.0.14

To reproduce, checkout with Klarna.
In Stores -> Config -> DotDigital -> Transactional SMS -> Everything Enabled No (we don't use it)
Go checkout.

After going through and checking out with Klarna - you'll see:
Dotdigitalgroup\Sms\Model\Queue\OrderItem\Data\AdditionalData::setOrderStatus(): Argument #1 ($orderStatus) must be of type string, null given, called in vendor/dotdigital/dotdigital-magento2-extension-sms/Model/Queue/OrderItem/NewOrder.php on line 28

Payment will have been captured, but no order in Magento.

Then another idea in the observers, there are no checks on the Enabled configs being done before attempting to do anything. 
Might also be a good idea to use your Settings and check if users even have enabled SMS at the start of the observers - Just my opinion if it's not enabled on a store why is any functional code allowed to execute. Surely first thing in the observers should be to check if it's enabled or not before doing anything else.

But found it happening on production, re-creatable on test using Klarna Sandbox.

**Comment by sta1r** [comment]
Hi again @TomJ1588 - yes the check you refer to is [here](https://github.com/dotmailer/dotmailer-magento2-extension-sms/blob/master/Model/Queue/OrderItem/OrderItemNotificationEnqueuer.php#L69-L71). However we are building an object first (without a check, [here](https://github.com/dotmailer/dotmailer-magento2-extension-sms/blob/master/Model/Queue/OrderItem/NewOrder.php#L28)) then checking prior to queuing it. 

We'll investigate this. 

I imagine there may be other places that break in our code with your Klarna extension if order status is null at `sales_order_save_after`. We'll have a scan of other cases.

**Comment by mpchadwick** [comment]
@TomJ1588 we might be running into this too. Is there a short term solution you could share?

**Comment by sta1r** [comment]
Folks, are you using Klarna Payments or Klarna Checkout? Klarna Checkout works in our tests, since it sets a default order status in the options.

<img width="782" alt="Screenshot 2023-11-15 at 17 01 18" src="https://github.com/dotmailer/dotmailer-magento2-extension-sms/assets/77671/5828caf9-4c78-4e5c-855a-e6d4a61feb5c">

In terms of ways to fix this, we're planning to release this changeset (or similar) middle of next week: https://github.com/dotmailer/dotmailer-magento2-extension-sms/compare/develop...AM/233435-klarna-checkout?expand=1. Shout if you have preferences / concerns.

**Comment by mpchadwick** [comment]
@sta1r The issue is with Klarna Payments. I've reviewed the patch and it doesn't seem like a true fix. You are just preventing some code from running in the case where SMS isn't turned on. The way I'm interpreting it this means Dotdigital SMS won't be compatible with Klarna payments, so if the merchant will only be able to use one feature or the other.

**Comment by sta1r** [comment]
@mpchadwick Thanks, will try and test with Klarna Payments. My change to make status nullable [here](https://github.com/dotmailer/dotmailer-magento2-extension-sms/blob/AM/233435-klarna-checkout/Model/Queue/OrderItem/Data/AdditionalData.php#L54) should mean the original reported error shouldn't happen, even if Transactional SMS is switched on.

**Comment by TomJ1588** [comment]
> @TomJ1588 we might be running into this too. Is there a short term solution you could share?

@mpchadwick Yes - as we are not using sms - our solution was to get rid of it.
It's that time of year where Klarna and Buy now pay later gets busy, and it was causing issue with too many orders.

Our easiest quickest solution was to get rid of it by:

Edit composer.json
```
 "replace": {
        "dotdigital/dotdigital-magento2-extension-sms": "*"
    },
```
Then run:

```
composer remove dotdigital/dotdigital-magento2-os-package
composer require dotdigital/dotdigital-magento2-os-package
```
That will remove everything then re-install DotDigital without the SMS module.

@sta1r - Yeah it's Klarna Payments not checkout.

**Comment by sta1r** [comment]
This ought to bypass the issue too:
```
bin/magento module:disable Dotdigitalgroup_Sms
bin/magento setup:upgrade
```

**Comment by andyjv** [comment]
Disregard— I misunderstood this issue to be a different one

**Comment by sta1r** [comment]
Fixed in v2.0.0 by e4d304621ba2193f03b976453b197281496e4460.
